pkg/agent/openflow/pipeline.go

```diff
@@ -847,6 +851,206 @@ func (f *featureService) snatConntrackFlows() []binding.Flow {
+		} else {
+			flowBuilder = flowBuilder.MatchProtocol(binding.ProtocolUDP)
+		}
+	default:
+		flowBuilder = flowBuilder.MatchIPProtocolValue(packet.IsIPv6, packet.IPProto)
+	}
+	if packet.IPProto == protocol.Type_TCP || packet.IPProto == protocol.Type_UDP {
+		if endpointPackets != nil && endpointPackets[0].DestinationPort != 0 {
+			flowBuilder = flowBuilder.MatchDstPort(endpointPackets[0].DestinationPort, nil)
+		} else if packet.DestinationPort != 0 {
+			flowBuilder = flowBuilder.MatchDstPort(packet.DestinationPort, nil)
+		}
+		if packet.SourcePort != 0 {
+			flowBuilder = flowBuilder.MatchSrcPort(packet.SourcePort, nil)
+		}
+	}
+
+	return flowBuilder
+}
+
+// flowsToCapture generates flows for packet capture. dataplaneTag is used as a mark for the target flow.
+func (f *featurePodConnectivity) flowsToCapture(dataplaneTag uint8,
+	ovsMetersAreSupported,
+	senderOnly bool,
+	receiverOnly bool,
+	packet *binding.Packet,
+	endpointPackets []binding.Packet,
+	ofPort uint32,
+	timeout uint16) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(cookie.PacketCapture).Raw()
+	var flows []binding.Flow
+	tag := uint32(dataplaneTag)
+	var flowBuilder binding.FlowBuilder
+	if !receiverOnly {
+		// if not receiverOnly, ofPort is inPort
+		if endpointPackets == nil {
+			flowBuilder = ConntrackStateTable.ofTable.BuildFlow(priorityHigh).
+				Cookie(cookieID).
+				MatchInPort(ofPort).
+				MatchCTStateTrk(true).
+				Action().LoadToRegField(PacketCaptureMark, tag).
+				SetHardTimeout(timeout).
+				Action().GotoStage(stagePreRouting)
+			if packet.DestinationIP != nil {
+				flowBuilder = flowBuilder.MatchDstIP(packet.DestinationIP)
+			}
+		} else {
+			// generate flows to endpoints.
+			for _, epPacket := range endpointPackets {
+				tmpFlowBuilder := ConntrackStateTable.ofTable.BuildFlow(priorityHigh).
+					Cookie(cookieID).
+					MatchInPort(ofPort).
+					MatchCTStateTrk(true).
+					Action().LoadToRegField(PacketCaptureMark, tag).
+					SetHardTimeout(timeout).
+					Action().GotoStage(stagePreRouting)
+				tmpFlowBuilder.MatchDstIP(epPacket.DestinationIP)
+				flow := matchTransportHeader(packet, tmpFlowBuilder, endpointPackets).Done()
+				flows = append(flows, flow)
+			}
+		}
+	} else {
+		flowBuilder = L2ForwardingCalcTable.ofTable.BuildFlow(priorityHigh).
+			Cookie(cookieID).
+			MatchCTStateTrk(true).
+			MatchDstMAC(packet.DestinationMAC).
+			Action().LoadToRegField(TargetOFPortField, ofPort).
+			Action().LoadRegMark(OutputToOFPortRegMark).
+			Action().LoadToRegField(PacketCaptureMark, tag).
+			SetHardTimeout(timeout).
+			Action().GotoStage(stageIngressSecurity)
+		if packet.SourceIP != nil {
+			flowBuilder = flowBuilder.MatchSrcIP(packet.SourceIP)
+		}
+	}
+
+	// for sender only case, capture the first tracked packet for svc.
+	if senderOnly {
+		for _, ipProtocol := range f.ipProtocols {
+			tmpFlowBuilder := ConntrackStateTable.ofTable.BuildFlow(priorityHigh).
```

Comment:
I don't get how this is determined as svc

Reply:
sry the comments is misleading. have moved this block  under the `enpoints != nil` block (svc case).